src/sparseml/transformers/finetune/session_mixin.py

```diff
@@ -0,0 +1,461 @@
+TRAINER_STATE_NAME = "trainer_state.json"
+
+
+class SessionManagerMixIn:
+    """
+    Mix-In class to extend the Hugging Face Trainer class to support SparseML recipes
+    for one-shot and finetuning flows.
+
+    :param model: PyTorch model to run training on
+    :param model_state_path: path to Pytorch model checkpoint or saved model
+    :param recipe: path to recipe file to apply during training
+    :param recipe_args: additional kwargs to use for evaluating recipe
+    :param metadata_args: additional kwargs for configuring training
+    :param data_args: kwargs for configuring dataset loading
+    :param teacher: optional teacher model to use for distillation
+    """
+
+    def __init__(
+        self,
+        model: Module,
+        model_state_path: str,
+        recipe: Optional[str],
+        recipe_args: Optional[Union[Dict[str, Any], str]] = None,
+        metadata_args: Optional[List[str]] = None,
+        data_args: Optional["DataTrainingArguments"] = None,  # noqa: F821
+        teacher: Optional[Union[Module, str]] = None,
+        **kwargs,
+    ):
+        # instantiate necessary state, like managers, so we can override args
+        self.model_state_path = str(model_state_path)
+        self.recipe = recipe
+        self.recipe_args = recipe_args
+        self.teacher = teacher
+
+        # parse training and metadata args
+        training_args = kwargs.get("args")
+        self.metadata = (
+            self._extract_metadata(
+                metadata_args=metadata_args,
+                training_args_dict=training_args.to_dict(),
+                data_args_dict=asdict(data_args) if data_args else {},
+            )
+            if training_args and metadata_args
+            else None
+        )
+
+        # setup logger and session
+        self.logger_manager = LoggerManager(log_python=False)
+        session_manager.create_session()
+
+        # call Trainer initialization
+        super().__init__(model=model, **kwargs)
+
+        # setup callbacks and loss
+        self.optim_callbacks = TrainingLoopCallbacks(self)
+        self.callback_handler.add_callback(self.optim_callbacks)
+        self.callback_disable_fp16 = DisableHalfPrecisionCallback(self)
+        self.callback_handler.add_callback(self.callback_disable_fp16)
+        self.criterion = torch.nn.CrossEntropyLoss()
+
+        model_signature = inspect.signature(model.forward)
+        self._model_signature_columns = list(model_signature.parameters.keys())
+
+    def initialize_session(self, epoch: float, checkpoint: Optional[str]):
+        """
+        Initialize the SparseSession from the specified epoch, evaluates the recipe
+        and initialized the modifiers for the training session
+
+        :param epoch: Epoch to initialize session from, usually 0 unless loading
+        from a checkpoint
+        :param checkpoint: Optional checkpoint to initialize from to continue training
+        """
+        orig_state_dict = self.model.state_dict()
```

Comment:
would be good if we can move the model loading functionality out of a trainer mixin so that it can be easily re-used for stuff like export